File: lib/features/visitor/models/visitor_access.dart
// Models cho Visitor Access & QR
class VisitorAccess {
  final String id;
  final String residentId;
  final String apartmentCode;
  final String visitorName;
  final String? visitorPhone;
  final String? visitorEmail;
  final DateTime visitDate;
  final String? visitTime; // "14:00" format
  final String? purpose;
  final String qrCode; // QR code string
  final String qrCodeUrl; // URL để generate QR
  final AccessStatus status;
  final DateTime? checkedInAt;
  final DateTime? checkedOutAt;
  final DateTime createdAt;
  final DateTime expiresAt;

  VisitorAccess({
    required this.id,
    required this.residentId,
    required this.apartmentCode,
    required this.visitorName,
    this.visitorPhone,
    this.visitorEmail,
    required this.visitDate,
    this.visitTime,
    this.purpose,
    required this.qrCode,
    required this.qrCodeUrl,
    this.status = AccessStatus.pending,
    this.checkedInAt,
    this.checkedOutAt,
    required this.createdAt,
    required this.expiresAt,
  });

  factory VisitorAccess.fromJson(Map<String, dynamic> json) {
    return VisitorAccess(
      id: json['id'] as String,
      residentId: json['residentId'] as String,
      apartmentCode: json['apartmentCode'] as String,
      visitorName: json['visitorName'] as String,
      visitorPhone: json['visitorPhone'] as String?,
      visitorEmail: json['visitorEmail'] as String?,
      visitDate: DateTime.parse(json['visitDate'] as String),
      visitTime: json['visitTime'] as String?,
      purpose: json['purpose'] as String?,
      qrCode: json['qrCode'] as String,
      qrCodeUrl: json['qrCodeUrl'] as String,
      status: AccessStatus.values.firstWhere(
        (e) => e.toString().split('.').last == json['status'],
        orElse: () => AccessStatus.pending,
      ),
      checkedInAt: json['checkedInAt'] != null
          ? DateTime.parse(json['checkedInAt'] as String)
          : null,
      checkedOutAt: json['checkedOutAt'] != null
          ? DateTime.parse(json['checkedOutAt'] as String)
          : null,
      createdAt: DateTime.parse(json['createdAt'] as String),
      expiresAt: DateTime.parse(json['expiresAt'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'residentId': residentId,
      'apartmentCode': apartmentCode,
      'visitorName': visitorName,
      'visitorPhone': visitorPhone,
      'visitorEmail': visitorEmail,
      'visitDate': visitDate.toIso8601String(),
      'visitTime': visitTime,
      'purpose': purpose,
      'qrCode': qrCode,
      'qrCodeUrl': qrCodeUrl,
      'status': status.toString().split('.').last,
      'checkedInAt': checkedInAt?.toIso8601String(),
      'checkedOutAt': checkedOutAt?.toIso8601String(),
      'createdAt': createdAt.toIso8601String(),
      'expiresAt': expiresAt.toIso8601String(),
    };
  }

  bool get isExpired => DateTime.now().isAfter(expiresAt);
  bool get canCheckIn => status == AccessStatus.pending && !isExpired;
}

enum AccessStatus {
  pending,
  checkedIn,
  checkedOut,
  expired,
  cancelled,
}

File: ICitizen/ICitizenAPI/ICitizen/Domain/VisitorAccess.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace ICitizen.Domain
{
    /// <summary>
    /// Entity cho Visitor Access - Quản lý khách vào tòa nhà bằng QR code
    /// </summary>
    public class VisitorAccess
    {
        [Key]
        public Guid Id { get; set; }

        [Required]
        public Guid ResidentId { get; set; }

        [ForeignKey("ResidentId")]
        public ResidentProfile? Resident { get; set; }

        [Required]
        [MaxLength(20)]
        public string ApartmentCode { get; set; } = string.Empty;

        [Required]
        [MaxLength(200)]
        public string VisitorName { get; set; } = string.Empty;

        [MaxLength(20)]
        public string? VisitorPhone { get; set; }

        [MaxLength(200)]
        public string? VisitorEmail { get; set; }

        [Required]
        public DateTime VisitDate { get; set; }

        [MaxLength(10)]
        public string? VisitTime { get; set; } // Format: "14:00"

        [MaxLength(500)]
        public string? Purpose { get; set; }

        [Required]
        [MaxLength(100)]
        public string QrCode { get; set; } = string.Empty; // Format: VISITOR_{timestamp}_{hash}

        [MaxLength(500)]
        public string? QrCodeUrl { get; set; }

        [Required]
        [MaxLength(20)]
        public string Status { get; set; } = "pending"; // pending, checkedIn, checkedOut, expired, cancelled

        public DateTime? CheckedInAt { get; set; }

        public DateTime? CheckedOutAt { get; set; }

        [Required]
        public DateTime CreatedAt { get; set; }

        [Required]
        public DateTime ExpiresAt { get; set; }
    }
}



